The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9870882271/job/27257591038) [(plain)](https://github.com/rust-lang/rust/commit/90ea75ba3d6a6463a7fce9d5a6111b297a21f2b3/checks/27257591038/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.57s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
---- compiler/rustc_mir_build/src/build/matches/mod.rs - build::matches::Builder<'a,'tcx>::bind_matched_candidate_for_guard (line 2341) stdout ----
error: unexpected `...`
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2344:19
  |
5 |     => feed(foo), ... }
  |                   |
  |                   not a valid pattern
  |                   help: for a rest pattern, use `..` instead of `...`


error: `match` arm with no body
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2344:19
  |
5 |     => feed(foo), ... }
  |                   ^^^- help: add a body after the pattern: `=> todo!(),`
error: `..` patterns are not allowed here
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2344:19
  |
  |
5 |     => feed(foo), ... }
  |
  = note: only allowed in tuple, tuple struct, and slice patterns

error[E0433]: failed to resolve: use of undeclared type `Foo`
---

error[E0425]: cannot find function `inspect` in this scope
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2343:22
  |
4 | match place { foo if inspect(foo)

error[E0425]: cannot find function `feed` in this scope
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2344:8
  |
  |
5 |     => feed(foo), ... }

error: aborting due to 6 previous errors

Some errors have detailed explanations: E0425, E0433.
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- compiler/rustc_mir_build/src/build/matches/mod.rs - build::matches::Builder<'a,'tcx>::bind_matched_candidate_for_guard (line 2365) stdout ----
error: unexpected `...`
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2368:47
  |
5 |     => { let tmp2 = &mut place; feed(tmp2) }, ... }
  |                                               |
  |                                               not a valid pattern
  |                                               help: for a rest pattern, use `..` instead of `...`


error[E0422]: cannot find struct, variant or union type `Foo` in this scope
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2367:15
  |
4 | match place { Foo { .. } if { let tmp1 = & &mut place; inspect(*tmp1) }

error: `match` arm with no body
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2368:47
  |
  |
5 |     => { let tmp2 = &mut place; feed(tmp2) }, ... }
  |                                               ^^^- help: add a body after the pattern: `=> todo!(),`
error: `..` patterns are not allowed here
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2368:47
  |
  |
5 |     => { let tmp2 = &mut place; feed(tmp2) }, ... }
  |
  = note: only allowed in tuple, tuple struct, and slice patterns

error[E0433]: failed to resolve: use of undeclared type `Foo`
---

error[E0425]: cannot find function `inspect` in this scope
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2367:56
  |
4 | match place { Foo { .. } if { let tmp1 = & &mut place; inspect(*tmp1) }

error[E0425]: cannot find function `feed` in this scope
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2368:33
  |
  |
5 |     => { let tmp2 = &mut place; feed(tmp2) }, ... }

error: aborting due to 7 previous errors

Some errors have detailed explanations: E0422, E0425, E0433.
Some errors have detailed explanations: E0422, E0425, E0433.
For more information about an error, try `rustc --explain E0422`.
Couldn't compile the test.
---- compiler/rustc_mir_build/src/build/matches/mod.rs - build::matches::Builder<'a,'tcx>::bind_matched_candidate_for_guard (line 2357) stdout ----
error: unexpected `...`
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2360:19
  |
5 |     => feed(foo), ... }
  |                   |
  |                   not a valid pattern
  |                   help: for a rest pattern, use `..` instead of `...`


error: `match` arm with no body
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2360:19
  |
5 |     => feed(foo), ... }
  |                   ^^^- help: add a body after the pattern: `=> todo!(),`
error: `..` patterns are not allowed here
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2360:19
  |
  |
5 |     => feed(foo), ... }
  |
  = note: only allowed in tuple, tuple struct, and slice patterns

error[E0433]: failed to resolve: use of undeclared type `Foo`
---

error[E0425]: cannot find function `inspect` in this scope
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2359:30
  |
4 | match place { ref mut foo if inspect(foo)

error[E0425]: cannot find function `feed` in this scope
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2360:8
  |
  |
5 |     => feed(foo), ... }

error: aborting due to 6 previous errors

Some errors have detailed explanations: E0425, E0433.
Some errors have detailed explanations: E0425, E0433.
For more information about an error, try `rustc --explain E0425`.
Couldn't compile the test.
---- compiler/rustc_mir_build/src/build/matches/mod.rs - build::matches::Builder<'a,'tcx>::bind_matched_candidate_for_guard (line 2349) stdout ----
error: unexpected `...`
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2352:42
  |
5 |     => { let tmp2 = place; feed(tmp2) }, ... }
  |                                          |
  |                                          not a valid pattern
  |                                          help: for a rest pattern, use `..` instead of `...`


error[E0422]: cannot find struct, variant or union type `Foo` in this scope
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2351:15
  |
4 | match place { Foo { .. } if { let tmp1 = &place; inspect(*tmp1) }

error: `match` arm with no body
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2352:42
  |
  |
5 |     => { let tmp2 = place; feed(tmp2) }, ... }
  |                                          ^^^- help: add a body after the pattern: `=> todo!(),`
error: `..` patterns are not allowed here
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2352:42
  |
  |
5 |     => { let tmp2 = place; feed(tmp2) }, ... }
  |
  = note: only allowed in tuple, tuple struct, and slice patterns

error[E0433]: failed to resolve: use of undeclared type `Foo`
---

error[E0425]: cannot find function `inspect` in this scope
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2351:50
  |
4 | match place { Foo { .. } if { let tmp1 = &place; inspect(*tmp1) }

error[E0425]: cannot find function `feed` in this scope
##[error] --> compiler/rustc_mir_build/src/build/matches/mod.rs:2352:28
  |
  |
5 |     => { let tmp2 = place; feed(tmp2) }, ... }

error: aborting due to 7 previous errors

Some errors have detailed explanations: E0422, E0425, E0433.
---
    compiler/rustc_mir_build/src/build/matches/mod.rs - build::matches::Builder<'a,'tcx>::bind_matched_candidate_for_guard (line 2349)

test result: FAILED. 12 passed; 4 failed; 11 ignored; 0 measured; 0 filtered out; finished in 243.81ms

error: doctest failed, to rerun pass `-p rustc_mir_build --doc`
  local time: Wed Jul 10 09:15:05 UTC 2024
  network time: Wed, 10 Jul 2024 09:15:05 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/compiler/rustc_mir_build/src/build/matches/mod.rs
+++ b/compiler/rustc_mir_build/src/build/matches/mod.rs
@@ -2177,92 +2177,15 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
 
         self.ascribe_types(block, ascriptions);
 
-        // rust-lang/rust#27282: The `autoref` business deserves some
-        // explanation here.
-        //
-        // The intent of the `autoref` flag is that when it is true,
-        // then any pattern bindings of type T will map to a `&T`
-        // within the context of the guard expression, but will
-        // continue to map to a `T` in the context of the arm body. To
-        // avoid surfacing this distinction in the user source code
-        // (which would be a severe change to the language and require
-        // far more revision to the compiler), when `autoref` is true,
-        // then any occurrence of the identifier in the guard
-        // expression will automatically get a deref op applied to it.
-        //
-        // So an input like:
-        //
-        // ```
-        // let place = Foo::new();
-        // match place { foo if inspect(foo)
-        //     => feed(foo), ... }
-        // ```
-        //
-        // will be treated as if it were really something like:
-        //
-        // ```
-        // let place = Foo::new();
-        // match place { Foo { .. } if { let tmp1 = &place; inspect(*tmp1) }
-        //     => { let tmp2 = place; feed(tmp2) }, ... }
-        // ```
-        //
-        // And an input like:
-        //
-        // ```
-        // let place = Foo::new();
-        // match place { ref mut foo if inspect(foo)
-        //     => feed(foo), ... }
-        // ```
-        //
-        // will be treated as if it were really something like:
-        //
-        // ```
-        // let place = Foo::new();
-        // match place { Foo { .. } if { let tmp1 = & &mut place; inspect(*tmp1) }
-        //     => { let tmp2 = &mut place; feed(tmp2) }, ... }
-        // ```
-        //
-        // In short, any pattern binding will always look like *some*
-        // kind of `&T` within the guard at least in terms of how the
-        // MIR-borrowck views it, and this will ensure that guard
-        // expressions cannot mutate their the match inputs via such
-        // bindings. (It also ensures that guard expressions can at
-        // most *copy* values from such bindings; non-Copy things
-        // cannot be moved via pattern bindings in guard expressions.)
-        //
-        // ----
-        //
-        // Implementation notes (under assumption `autoref` is true).
-        //
-        // To encode the distinction above, we must inject the
-        // temporaries `tmp1` and `tmp2`.
-        //
-        // There are two cases of interest: binding by-value, and binding by-ref.
-        //
-        // 1. Binding by-value: Things are simple.
-        //
-        //    * Establishing `tmp1` creates a reference into the
-        //      matched place. This code is emitted by
-        //      bind_matched_candidate_for_guard.
-        //
-        //    * `tmp2` is only initialized "lazily", after we have
-        //      checked the guard. Thus, the code that can trigger
-        //      moves out of the candidate can only fire after the
-        //      guard evaluated to true. This initialization code is
-        //      emitted by bind_matched_candidate_for_arm.
-        //
-        // 2. Binding by-reference: Things are tricky.
-        //
-        //    * Here, the guard expression wants a `&&` or `&&mut`
-        //      into the original input. This means we need to borrow
-        //      the reference that we create for the arm.
-        //    * So we eagerly create the reference for the arm and then take a
-        //      reference to that.
+        // Lower an instance of the arm guard (if present) for this candidate,
+        // and then perform bindings for the arm body.
         if let Some((arm, match_scope)) = arm_match_scope
             && let Some(guard) = arm.guard
         {
             let tcx = self.tcx;
 
+            // Bindings for guards require some extra handling to automatically
+            // insert implicit references/dereferences.
             self.bind_matched_candidate_for_guard(block, schedule_drops, bindings.clone());
             let guard_frame = GuardFrame {
                 locals: bindings.clone().map(|b| GuardFrameLocal::new(b.var_id)).collect(),
@@ -2402,6 +2325,82 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         }
     }
 
+    /// Binding for guards is a bit different from binding for the arm body,
+    /// because an extra layer of implicit reference/dereference is added.
+    ///
+    /// The idea is that any pattern bindings of type T will map to a `&T` within
+    /// the context of the guard expression, but will continue to map to a `T`
+    /// in the context of the arm body. To avoid surfacing this distinction in
+    /// the user source code (which would be a severe change to the language and
+    /// require far more revision to the compiler), any occurrence of the
+    /// identifier in the guard expression will automatically get a deref op
+    /// applied to it. (See the caller of [`Self::is_bound_var_in_guard`].)
+    ///
+    /// So an input like:
+    ///
+    /// ```ignore (illustrative)
+    /// let place = Foo::new();
+    /// match place { foo if inspect(foo)
+    ///     => feed(foo), ... }
+    /// ```
+    ///
+    /// will be treated as if it were really something like:
+    ///
+    /// ```ignore (illustrative)
+    /// let place = Foo::new();
+    /// match place { Foo { .. } if { let tmp1 = &place; inspect(*tmp1) }
+    ///     => { let tmp2 = place; feed(tmp2) }, ... }
+    /// ```
+    ///
+    /// And an input like:
+    ///
+    /// ```ignore (illustrative)
+    /// let place = Foo::new();
+    /// match place { ref mut foo if inspect(foo)
+    ///     => feed(foo), ... }
+    /// ```
+    ///
+    /// will be treated as if it were really something like:
+    ///
+    /// ```ignore (illustrative)
+    /// let place = Foo::new();
+    /// match place { Foo { .. } if { let tmp1 = & &mut place; inspect(*tmp1) }
+    ///     => { let tmp2 = &mut place; feed(tmp2) }, ... }
+    /// ```
+    /// ---
+    ///
+    /// ## Implementation notes
+    ///
+    /// To encode the distinction above, we must inject the
+    /// temporaries `tmp1` and `tmp2`.
+    ///
+    /// There are two cases of interest: binding by-value, and binding by-ref.
+    ///
+    /// 1. Binding by-value: Things are simple.
+    ///
+    ///    * Establishing `tmp1` creates a reference into the
+    ///      matched place. This code is emitted by
+    ///      [`Self::bind_matched_candidate_for_guard`].
+    ///
+    ///    * `tmp2` is only initialized "lazily", after we have
+    ///      checked the guard. Thus, the code that can trigger
+    ///      moves out of the candidate can only fire after the
+    ///      guard evaluated to true. This initialization code is
+    ///      emitted by [`Self::bind_matched_candidate_for_arm_body`].
+    ///
+    /// 2. Binding by-reference: Things are tricky.
+    ///
+    ///    * Here, the guard expression wants a `&&` or `&&mut`
+    ///      into the original input. This means we need to borrow
+    ///      the reference that we create for the arm.
+    ///    * So we eagerly create the reference for the arm and then take a
+    ///      reference to that.
+    ///
+    /// ---
+    ///
+    /// See these PRs for some historical context:
+    /// - <https://github.com/rust-lang/rust/pull/49870> (introduction of autoref)
+    /// - <https://github.com/rust-lang/rust/pull/59114> (always use autoref)
     fn bind_matched_candidate_for_guard<'b>(
         &mut self,
         block: BasicBlock,
@@ -2433,10 +2432,13 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
             );
             match binding.binding_mode.0 {
                 ByRef::No => {
+                    // The arm binding will be by value, so for the guard binding
+                    // just take a shared reference to the matched place.
                     let rvalue = Rvalue::Ref(re_erased, BorrowKind::Shared, binding.source);
                     self.cfg.push_assign(block, source_info, ref_for_guard, rvalue);
                 }
                 ByRef::Yes(mutbl) => {
+                    // The arm binding will be by reference, so eagerly create it now.
                     let value_for_arm = self.storage_live_binding(
                         block,
                         binding.var_id,
@@ -2448,6 +2450,7 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                     let rvalue =
                         Rvalue::Ref(re_erased, util::ref_pat_borrow_kind(mutbl), binding.source);
                     self.cfg.push_assign(block, source_info, value_for_arm, rvalue);
+                    // For the guard binding, take a shared reference to that reference.
                     let rvalue = Rvalue::Ref(re_erased, BorrowKind::Shared, value_for_arm);
                     self.cfg.push_assign(block, source_info, ref_for_guard, rvalue);
                 }

